File: love/card.lua
-- Blackjack
-- Copyright 2015 Sam Saint-Pettersen
-- Released under the MIT/X11 License.
--
-- Powered by the LÖVE Game Engine

--- Card class for Blackjack.
-- @copyright 2015 Sam Saint-Pettersen

Card = {}
Card.__index = Card

--- Get an image path from card string pattern.
-- @param [string] card String representation for card.
-- @return [string] Path to card graphic.
function Card_getImage(card)
	local suit = ''
	if card:match('(h)') ~= nil then
		suit = 'h'
	elseif card:match('(d)') ~= nil then
		suit = 'd'
	elseif card:match('(c)') ~= nil then
		suit = 'c'
	elseif card:match('(s)') ~= nil then
		suit = 's'
	end
	local rank = card:match('%[*(%d*%u*)%]*')
	return string.format('gfx/%s%s.png', suit, rank)
end

--- Card represents a single playing card.
-- @constructor
-- @param [string] card Path for card graphic.
-- @param [number] posX X position for card.
-- @param [number] posY Y position for card.
function Card.create(card, posX, posY)
	local self = setmetatable({}, Card)
	self.src = card
	self.image = love.graphics.newImage(card)
	self.posX = posX
	self.posY = posY

	return self
end

--- Set X, Y position for card.
-- @param [number] posX X position for card.
-- @param [number] posY Y position for card.
function Card:setXY(posX, posY)
	self.posX = posX
	self.posY = posY
end

--- Get X, Y position of card.
-- @return [number, number] X, Y position of card.
function Card:getXY()
	return self.posX, self.posY
end

--- Get source for image.
-- @return [string] Image source (path). 
function Card:getImageSrc()
	return self.src
end

--- Draw the card.
function Card:draw()
	love.graphics.draw(self.image, self.posX, self.posY)
end

File: html5/enchant.js_ts/card.ts
/*
	Blackjack
	Copyright 2015 Sam Saint-Pettersen
	Released under the MIT/X11 License.
	
	HTML5/enchant.js TypeScript implementation
*/

/// <reference path="enchant.d.ts" />
/// <reference path="graphics.ts" />

class Card {
	private sprite: Sprite;
	private src: string;

	/**
	 * Card represents a single playing card.
	 * @constructor
	 * @param {string} card - Data URI for card graphic.
	 * @param {number} posX - X position for card.
	 * @param {number} posY - Y position for card.
	 * @param {Core} game - Enchant.js game object.
	*/
	constructor(card: string, posX: number, posY: number, game: Core) {
		this.sprite = new Sprite(71, 96);
		this.sprite.image = game.assets[card];
		this.sprite.x = posX;
		this.sprite.y = posY;
		this.src = card;
	}

	/** 
	 * Get image path from card string pattern.
	 * @static
	 * @param {string} card - Card string pattern.
	 * @returns {string} Path for card image.
	*/
	public static getImage(card: string) {
		if(card == 'c' || card == 'd')
			return g + card + '.png';

		var suit: string = '';
		if (new RegExp('(h)').test(card))
			suit = 'h';

		else if (new RegExp('(d)').test(card))
			suit = 'd';

		else if (new RegExp('(c)').test(card))
			suit = 'c';

		else if (new RegExp('(s)').test(card))
			suit = 's';

		var rank: string = card.match(/\[*([0-9A-Z]*)\]*/)[1];
		return "${g}${suit}${rank}.png";
	}

	/**
	 * Get source for image.
	 * @returns {string} Image source.
	*/
	public getImageSrc(): string {
		return this.src;
	}

	/**
	 * Set X, Y position for card.
	 * @param {number} posX - X position for card.
	 * @param {number} posY - Y position for card.
	*/
	public setXY(posX: number, posY: number): void {
		this.sprite.x = posX;
		this.sprite.y = posY;
	}

	/**
	 * Get X, Y position for card.
	 * @returns {number[]} X, Y position of card.
	*/
	public getXY(): number[] {
		return [this.sprite.x, this.sprite.y];
	}

	/**
	 * Draw the card.
	*/
	public draw(): Sprite {
		return this.sprite;
	}
}
